### PR TITLE
Exclude unassigned with code action

### PR DESCRIPTION
**Pull request by angelozerr**
Exclude unassigned with code action

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR adds the capability to exclude the error validation for unassigned config property inside Java file.

See the following demo:

CodeActionExcludeUnassignedDemo

Please note that once code action is executed you must modify the Java file in order to the error validation disapear. This limitation is blocked by https://github.com/eclipse/lsp4mp/issues/163

The command insert thesettings in Workspace level with

```json
"microprofile.tools.validation.unassigned.excluded": [
		"FROM_ENV"
	]
```

**Comment by AlexXuChen** [comment]
The diagnostic only updates on file save, I feel it should update on CodeAction trigger as well.

Additionally, if I remove the excluded members in `microprofile.tools.validation.unassigned.excluded`, the diagnostic doesn't update when I reopen the file, which I believe is beneficial to support.
Peek 2021-09-27 11-11

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/NoValueAssignedToPropertyQuickFix.java [hunk 1]
```diff
@@ -117,4 +126,30 @@ private static String getPropertyName(Diagnostic diagnostic, JavaCodeActionConte
...
+	 * configuration
+	 *
+	 * @param item       the item to add to the client configuration array
+	 * @param diagnostic the diagnostic for the <code>CodeAction</code>
+	 * @return a code action that causes <code>item</code> to be added to
+	 *         <code>microprofile.tools.validation.unknown.excluded</code> client
+	 *         configuration
+	 */
+	private CodeAction createAddToExcludedCodeAction(String item, Diagnostic diagnostic) {
+		CodeAction insertCodeAction = new CodeAction("Exclude '" + item + "' from unassigned property validation?");
```
Comment (AlexXuChen) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/NoValueAssignedToPropertyQuickFix.java:143]: Use `String.format("Exclude '%s' " from unassigned property validation)` (question mark unnecessary)

Comment (AlexXuChen) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/NoValueAssignedToPropertyQuickFix.java:143]: Use `String.format("Exclude '%s' " from unassigned property validation", item)` (question mark unnecessary)

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/NoValueAssignedToPropertyQuickFix.java [hunk 2]
```diff
@@ -117,4 +126,30 @@ private static String getPropertyName(Diagnostic diagnostic, JavaCodeActionConte
...
+
+		ConfigurationItemEdit configItemEdit = new ConfigurationItemEdit(
+				"microprofile.tools.validation.unassigned.excluded", ConfigurationItemEditType.add, item);
+
+		Command command = new Command("Add " + item + " to unknown excluded array",
```
Comment (AlexXuChen) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/NoValueAssignedToPropertyQuickFix.java:148]: Use `String.format("Add %s to unknown excluded array", item)`

**Review by AlexXuChen — CHANGES_REQUESTED** [review]

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/NoValueAssignedToPropertyQuickFix.java (hunk 1 above)
Comment (angelozerr) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/NoValueAssignedToPropertyQuickFix.java:143]: Good catch i will fix it

**Comment by angelozerr** [comment]
@AlexXuChen all limitation that you find comes from that settings are not tracked on jdt side. Once https://github.com/eclipse/lsp4mp/issues/16 will be implemented it should fix those limitation.

**Comment by AlexXuChen** [comment]
> @AlexXuChen all limitation that you find comes from that settings are not tracked on jdt side. Once #16 will be implemented it should fix those limitation.

Okay, with that in mind then, everything is working properly

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/NoValueAssignedToPropertyQuickFix.java (hunk 1 above)
Comment (rgrunber) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/NoValueAssignedToPropertyQuickFix.java:143]: I would maybe call it just "Exclude '%s' from property validation". The unassigned part might be a bit confusing to users.

**Comment by rgrunber** [comment]
Feel free to merge once the needed changes are made.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/NoValueAssignedToPropertyQuickFix.java (hunk 1 above)
Comment (angelozerr) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/NoValueAssignedToPropertyQuickFix.java]: fixed, now I use MessageFormat

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/NoValueAssignedToPropertyQuickFix.java (hunk 2 above)
Comment (angelozerr) [reply microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/NoValueAssignedToPropertyQuickFix.java]: fixed

**Comment by angelozerr** [comment]
> The diagnostic only updates on file save, I feel it should update on CodeAction trigger as well.

> Additionally, if I remove the excluded members in microprofile.tools.validation.unassigned.excluded, the diagnostic doesn't update when I reopen the file, which I believe is beneficial to support.

Sorry with my last comment, to fix that, it's not #16. I have fixed that in the PR and now when you execute code action, the diagnostics is refreshed. If you update your settings.json the diagnostics is refreshed too.

**Comment by angelozerr** [comment]
Just for your info, I intreoduced MicroProfileCodeActionFactory which is a code action factory specific for MicroProfile and which is used in properties and java file both. It's the reason why I create this MicroProfileCodeActionFactory in the commons package.

**Pull request merged**